Draft implementation of topological ordering

garyanaplan opened the pull request:
I've tried to implement this in a style that is sympathetic with the existing style in the Dfs/Bfs implementations.

Very happy to make changes or accept suggestions for improvements.

garyanaplan committed:
Draft implementation of topological ordering

octavonce commented on src/graph.rs:
```diff
@@ -974,6 +974,43 @@ impl<T> Graph<T> {
+    /// of the graph which follows a DFS based
+    /// topological order (Kahn's algorithm).
+    ///
+    /// Topological sorting is not possible for
```
```
Topological sorting is not possible for
graphs which contain a cycle. You may
use topo.is_cylic() == false to verify
that your graph is a DAG.
```

What happens if the user doesn't check this? Maybe we should add a panic?

garyanaplan replied:
You are right, that's a problem.

I've pushed a fix and added a test to exercise the panic code. I can't figure out how to make the test check for PANIC_MSG, so I hard-coded the text into the test. That's a bit annoying.

octavonce commented on src/iterators/topo.rs:
```diff
@@ -0,0 +1,156 @@
+    iterable: &'a Graph<T>,
+    /// Processed vertices
+    vertices: Vec<&'a VertexId>,
+    /// Working set of vertoces
```
Doc typo

octavonce reviewed (commented):
Just two minor things but besides those, it looks good!

garyanaplan committed:
- (REVIEW) Fix Typo.

garyanaplan committed:
- Improve robustness of cycle detection

process_vertex() now will panic (except for calls from is_cyclic()) if
the number of vertices seen don't match the graph's vertex count when
processing is complete.

Annoyingly, I can't figure out how to make the expected test for the
panic take a const value as an argument, so I've copy/pasted the string
into the test macro text.

garyanaplan committed:
- Improve documentation

Attempt to explain when a panic!() may be hit in use.

garyanaplan commented:
Ok I think the documentation change and the addition of the panic!() is all I want to do on this right now.

octavonce reviewed (approved):
Ok, this looks great! Merging now.
